### PR TITLE
Display data returned in `apply_ufunc` error message

### PR DESCRIPTION
**Pull request by max-sixty**
This makes debugging much easier!

- [x] Tests added

### CODE REVIEW

File: xarray/core/computation.py [hunk 1]
```diff
@@ -784,21 +782,21 @@ def func(*arrays):
...
-                    "size of dimension {!r} on inputs was unexpectedly "
-                    "changed by applied function from {} to {}. Only "
+                    f"size of dimension '{dim}' on inputs was unexpectedly "
+                    f"changed by applied function from {dim_sizes[dim]} to {new_size}. Only "
                     "dimensions specified in ``exclude_dims`` with "
-                    "xarray.apply_ufunc are allowed to change size.".format(
-                        dim, dim_sizes[dim], new_size
-                    )
+                    "xarray.apply_ufunc are allowed to change size. "
+                    "The data returned was:\n\n{data!r}"
```
Comment (Illviljan) [inline xarray/core/computation.py:799]: Did you try the error with a large sized data? It still looks good?

Comment (max-sixty) [reply xarray/core/computation.py]: Good point. I've now wrapped them in `short_array_repr`... Or lmk if another approach is better

**Review by headtr1ck — APPROVED** [review]
Seems helpful! Thanks

+1 for typing improvements

File: xarray/core/computation.py [hunk 2]
```diff
@@ -766,11 +767,10 @@ def func(*arrays):
...
     ):
         raise ValueError(
-            "applied function does not have the number of "
-            "outputs specified in the ufunc signature. "
-            "Result is not a tuple of {} elements: {!r}".format(
-                signature.num_outputs, result_data
-            )
+            f"applied function does not have the number of "
+            f"outputs specified in the ufunc signature. "
+            f"Result is not a tuple of {signature.num_outputs} elements:\n\n"
```
Comment (dcherian) [inline xarray/core/computation.py:772]: Does this work well if we receive a list of arrays?

Perhaps we should add 
```python
f"Received a {type(result_data)} with {len(result_data)} elements instead."
```

**Review by dcherian — COMMENTED** [review]
Very nice, I had one comment.

(sorry for the late review)

File: xarray/core/computation.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/computation.py:772]: Ah, good point. #8192

**Pull request merged**